api/client/copy.go

```diff
@@ -0,0 +1,218 @@
+		srcInfo := strings.SplitN(srcPath, ":", 2)
+		srcContainer, srcPath = srcInfo[0], srcInfo[1]
+	}
+
+	if strings.Contains(dstPath, ":") {
+		// Copy to a container.
+		direction |= toContainer
+		dstInfo := strings.SplitN(dstPath, ":", 2)
+		dstContainer, dstPath = dstInfo[0], dstInfo[1]
+	}
+
+	switch direction {
+	case fromContainer:
+		// Copy from a container to a local path.
+		return cli.copyFromContainer(srcContainer, srcPath, dstPath)
+	case toContainer:
+		// Copy from a local path to a container.
+		return cli.copyToContainer(srcPath, dstContainer, dstPath)
+	case acrossContainers:
+		// Copy from one container to another.
+		return cli.copyAcrossContainers(srcContainer, srcPath, dstContainer, dstPath)
+	default:
+		// User didn't specify any container.
+		return fmt.Errorf("error: Wrong path format. Must specify a source and/or destination container")
+	}
+}
+
+type containerPathStat struct {
+	Name    string      `json:"name"`
+	AbsPath string      `json:"absPath"`
+	Size    int64       `json:"size"`
+	Mode    os.FileMode `json:"mode"`
+	Mtime   time.Time   `json:"mtime"`
+}
+
+func (cli *DockerCli) statContainerPath(containerName, path string) (containerPathStat, error) {
+	var stat containerPathStat
+
+	query := make(url.Values, 1)
+	query.Set("path", path)
+
+	urlStr := fmt.Sprintf("/containers/%s/stat-path?%s", containerName, query.Encode())
+
+	response, statusCode, err := cli.call("GET", urlStr, nil, false)
+	if err != nil {
+		return stat, err
+	}
+
+	defer response.Close()
+
+	if statusCode != 200 {
+		return stat, fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	err = json.NewDecoder(response).Decode(&stat)
+
+	return stat, err
+}
+
+func resolveLocalPath(localPath string) (absPath string, err error) {
+	if absPath, err = filepath.Abs(localPath); err != nil {
+		return
+	}
+
+	return archive.PreserveTrailingDotOrSeparator(absPath, localPath), nil
+}
+
+func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string) error {
+	// Get an absolute destination path.
+	dstPath, err := resolveLocalPath(dstPath)
+	if err != nil {
+		return err
+	}
+
+	// Start by stat-ing the source path in the container.
+	srcStat, err := cli.statContainerPath(srcContainer, srcPath)
+	if err != nil {
+		// Some error, like container or resource does not
+		// exist, so we can't perform the copy operation.
+		return err
+	}
+
+	query := make(url.Values, 1)
+	query.Set("path", srcPath)
+
+	urlStr := fmt.Sprintf("/containers/%s/archive-path?%s", srcContainer, query.Encode())
+
+	response, statusCode, err := cli.call("GET", urlStr, nil, false)
+	if err != nil {
+		return err
+	}
+
+	defer response.Close()
+
+	if statusCode != 200 {
+		return fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	// Prepare source copy info.
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  srcStat.Mode.IsDir(),
+	}
+
+	return archive.CopyTo(response, srcInfo, dstPath)
```

Comment:
> will this deal with all cases of:
> ...

It should, take a look at this part of the patch: https://github.com/jlhawn/docker/commit/5dd311779804641165a4332464e88ebb7cd8f388#diff-d50c86374d8a7d3552ed2ae1b708a7d4R157

> If we're doing a single file copy, is there any reason why we're not just sending the raw file instead of tar-ing it up?

Using tar archives of the content allows for a more generic implementation since the archive can container either a file or a directory of files. Plus, the archive entry headers hold all of the metadata for the content too, so that's a plus I think.